### ISSUE TITLE
Weewx 4.x aka python 3

### ISSUE THREAD

**Issue by silasmoeckel** [AussieSusan/HP1000#5]
Any plans to update this for the 4.x weewx versions?  Looks like a few hiccups related to python3 defaulting to UTF-8 (line 323 for starters).

**Comment by W2NJL** [comment]
I was able to install the HP1000 extension with python3, but the HP1000 driver does not appear in the list re-running the weewx configuration process.  I suspect this may be related to what you are describing?

**Comment by siblake7** [comment]
Ditto
Has anyone converted this driver to work with python3.  I'm using python3.5.3 and I get the error below.  I have uninstalled and re-installed HP1000.  I also get the issue that when I reconfigure "sudo python ./bin/wee_config --reconfigure"  the HP1000 driver is not listed.  I assume because it isn't compiled successfully after the install

Sep  5 10:34:30 raspberrypi weewx[2597] CRITICAL __main__: Caught unrecoverable exception:
Sep  5 10:34:30 raspberrypi weewx[2597] CRITICAL __main__:     ****  invalid syntax (HP1000.py, line 243)
Sep  5 10:34:30 raspberrypi weewx[2597] CRITICAL __main__:     ****  Traceback (most recent call last):
Sep  5 10:34:30 raspberrypi weewx[2597] CRITICAL __main__:     ****    File "/usr/share/weewx/weewxd", line 148, in main
Sep  5 10:34:30 raspberrypi weewx[2597] CRITICAL __main__:     ****      engine = weewx.engine.StdEngine(config_dict)
Sep  5 10:34:30 raspberrypi weewx[2597] CRITICAL __main__:     ****    File "/usr/share/weewx/weewx/engine.py", line 69, in __init__
Sep  5 10:34:30 raspberrypi weewx[2597] CRITICAL __main__:     ****      self.setupStation(config_dict)
Sep  5 10:34:30 raspberrypi weewx[2597] CRITICAL __main__:     ****    File "/usr/share/weewx/weewx/engine.py", line 92, in setupStation
Sep  5 10:34:30 raspberrypi weewx[2597] CRITICAL __main__:     ****      __import__(driver)
Sep  5 10:34:30 raspberrypi weewx[2597] CRITICAL __main__:     ****    File "/usr/share/weewx/user/HP1000.py", line 243
Sep  5 10:34:30 raspberrypi weewx[2597] CRITICAL __main__:     ****      except KeyError, e:
Sep  5 10:34:30 raspberrypi weewx[2597] CRITICAL __main__:     ****                     ^
Sep  5 10:34:30 raspberrypi weewx[2597] CRITICAL __main__:     ****  SyntaxError: invalid syntax
Sep  5 10:34:30 raspberrypi weewx[2597] CRITICAL __main__:     ****  Exiting.

**Comment by PoloG12** [comment]
Hi all,
I was in contact with Susan that made this great driver and she confirm that the driver is not compatible with Python 3. She can't update the driver so I hope that someone with a good level in Python could update the script.
Anyway, I already found some errors that anybody can change:

Replace the line (around 244):
except KeyError:
by:
except KeyError as e:

Line (around 324)
sock.sendto(bcData, (self.ip_address_mask, UDP_BROADCAST_PORT))
by: sock.sendto(bcData.encode(), (UDP_BROADCAST_IP, UDP_BROADCAST_PORT))

Line (around 438):
self.ws_socket.send(self.create_cmd_string(argument='SETUP'))
by:
self.ws_socket.send(self.create_cmd_string(argument='SETUP').encode())

It doesn't fix the general issue but it's a step. Now the error is a (classic?) Python 3 error:
For the command: cmd_packet = struct.pack('8s8s16s2i', 'PC2000', 'READ', 'HISTORY_FILE', 40, 0)
Error: struct.error: argument for 's' must be a bytes object

All the .pack() methods must be changed. If someone knows about Python, it will help us a lot.

Regards.

**Comment by Damien0505** [comment]
I started trying to convert it the other week, but I had other things I had to get back too.
I have cheated for the moment, having the odler version of weewx with Python 2.7 running on a Raspberry PI and connecting to the database on the machine I had weewx running on previously.  It grabs data from the HP1000, still writes it to my original DB and files are written to the webserver via NFS.  

Hopefully AussieSusan maybe able to update it when she gets a chance.  :)
